test/session-pool.js

```diff
@@ -16,2022 +16,1575 @@
 
 'use strict';
 
-var assert = require('assert');
-var common = require('@google-cloud/common');
-var delay = require('delay');
-var events = require('events');
-var extend = require('extend');
-var PQueue = require('p-queue');
-var proxyquire = require('proxyquire');
-var stackTrace = require('stack-trace');
-var through = require('through2');
-var timeSpan = require('time-span');
-
-var pQueueOverride = null;
-function FakePQueue(options) {
-  return new (pQueueOverride || PQueue)(options);
-}
-
-var fakeStackTrace = extend({}, stackTrace);
+const assert = require('assert');
+const common = require('@google-cloud/common');
+const extend = require('extend');
+const through = require('through2');
 
 describe('SessionPool', function() {
-  var SessionPool;
-  var sessionPool;
+  let SessionPool;
 
-  var DATABASE = {
+  let DATABASE = {
     request: common.util.noop,
     requestStream: common.util.noop,
   };
 
   before(function() {
-    SessionPool = proxyquire('../src/session-pool.js', {
-      'p-queue': FakePQueue,
-      'stack-trace': fakeStackTrace,
-    });
-  });
-
-  beforeEach(function() {
-    sessionPool = new SessionPool(DATABASE);
-    sessionPool.onClose_ = new Promise(function() {});
+    SessionPool = require('../src/session-pool.js');
   });
 
-  afterEach(function() {
-    pQueueOverride = null;
-    fakeStackTrace.get = stackTrace.get;
-  });
+  describe('tests with no min sessions sessionPool before hook', function() {
+    let sessionPool;
+    // Because we create a sessionPool here for every test tests which want to open
+    // their own pool must first close this one and wait on its returned promise.
+    // Otherwise the tests will have resource leaks.
+    beforeEach(function() {
+      sessionPool = new SessionPool(DATABASE, {
+        maxReads: 10,
+        maxWrites: 5,
+        acquireTimeout: 50,
+      });
+    });
 
-  describe('instantiation', function() {
-    it('should localize the database instance', function() {
-      assert.strictEqual(sessionPool.database, DATABASE);
+    afterEach(function() {
+      clearTimeout(sessionPool.pingTimeoutHandle);
```

Comment:
Shouldn't this be handled by the call to `close()`?

Reply:
This is fixed